Motion control node

kylestach opened the pull request:
## Description
Clean up the motion control node in preparation for moving to a pub-sub.

## Associated Issue
Issue #1424 

## Design Documents
[ROS Roadmap](https://docs.google.com/document/d/1wix-ZK9NPBY0zC8T9LyUihT1ONubXMKxHvb8KYmJkNQ/edit#)

## Steps to test
Motion control should experience no regressions. Run adaptive formation to test.

kylestach committed:
Clean up motion control node and surrounding utilities

kylestach committed:
Make pretty-lines

JNeiger commented on soccer/motion/MotionControl.cpp:
```diff
@@ -35,244 +30,150 @@ void MotionControl::createConfiguration(Configuration* cfg) {
+        } else if (maybe_target) {
+            _drawer->drawCircle(maybe_target->motion.pos, .15, Qt::green,
+                                "Planning");
+        }
```
You lost the velocity debug drawer

kylestach replied:
Fixed

JNeiger commented on soccer/motion/MotionControl.cpp:
```diff
@@ -35,244 +30,150 @@ void MotionControl::createConfiguration(Configuration* cfg) {
-        }
-    }
-    //}
+    // TODO: Reset and return if robot is joystick-controlled.
```
I would make sure this is added before actually merging. It would be a real surprise to whoever tests next if you dont lol

kylestach replied:
Yeah, I'll add that in tomorrow. I just couldn't think of a good place for the "is joystick controlled" field to go in Context.

JNeiger commented on soccer/motion/MotionControlNode.cpp:
```diff
@@ -1,28 +1,30 @@
+    bool force_stop = game_state.state == GameState::State::Halt;
+    for (int i = 0; i < Num_Shells; i++) {
+        MotionControl& controller = _controllers[i];
+        MotionSetpoint* setpoint = &(*setpoints)[i];
```
This looks a little jank from a syntax pov, but there's not really a way to fix it :/

kylestach replied:
I mean, it's not really that weird. It's "address of the value at index i in the array pointed to by setpoints", which is a pretty reasonable thing to express. Either way not much we can do about it.

kylestach committed:
Make ConfigItem non-copyable/movable, save RobotConfig in ~Processor

Normally we don't care what happens after ~Processor, but in the test
system (specifically for Python code) Processor is instantiated and
killed multiple times. That means that config variables (which get
initialized before main()) need to be saved and restored at the
beginning and end of Processor's lifespan.

JNeiger commented on logging/CMakeLists.txt:
```diff
@@ -8,3 +7,7 @@ target_link_libraries(simple_logger common rc-fshare)
+
+# We need XML because Context includes configurables,
+# which uses XML to represent the tree.
+qt5_use_modules(log_converter Xml Core)
```
This feels unrelated to this pr. Is this part of the process to save context every frame as a data logging solution or something?

kylestach replied:
No, this is because the per-robot configurables got added into context, which log_converter depends upon. When this isn't here log_converter will (eventually) include context, which will include some Qt files from the Xml package, so it will fail. The "proper" fix is that `qt5_use_modules` should transitively allow the include directories (the macro is kind of broken, but it's not our code) - however, this is a reasonable workaround.

JNeiger commented on soccer/RobotConfig.hpp:
```diff
@@ -78,11 +84,14 @@ class RobotConfig {
-    ConfigBool* ball_sense_enabled;
-    ConfigBool* dribbler_enabled;
+    RobotStatus(RobotStatus&&) = default;
+    RobotStatus& operator=(RobotStatus&&) = default;
```
Are we sure we want to allow copies with all the unique ptrs here?

kylestach replied:
That's a move. Moving is okay, copying isn't.

JNeiger replied:
Oof, super misread that

JNeiger commented on soccer/motion/MotionControl.cpp:
```diff
@@ -35,244 +30,150 @@ void MotionControl::createConfiguration(Configuration* cfg) {
-    target.vel = _lastWorldVelCmd + targetAccel * dt.count();
-
-    _lastWorldVelCmd = target.vel;
-    _lastCmdTime = RJ::now();
```
Just to throw this out there, it's more correct without all the motion control hacks, so removing them is the "correct" thing to do, but it did bite us in the ass when I did that on firmware.  It might be prudent to leave them in until tuning phase, and then remove them if they're not needed.

kylestach replied:
That's true. However, the minimum-velocity hacks actually significantly (negatively) impact sim performance, and so I think they belong on the robot side of motion control regardless. I'd honestly rather optimize this part for sim anyways, considering that we are who knows how far from seeing an actual robot, and the first thing we do with new robots is motion control work where I'd rather have a clean(er) slate to work with anyway.

Acceleration _feedthrough_ is something that I think probably does belong here. I'll add it in tomorrow. Acceleration limiting is more questionable  - there's an argument to be said that it belongs on the robot, or that it belongs here, but either way I don't think it will be making much difference considering the motion planning already acceleration limits the goal (so it only matters here if feedback is large).

kylestach committed:
Add debug drawing for velocity in motion control

kylestach committed:
Move is_joystick_controlled to Context

 - This allows us to properly use it in motion control

kylestach committed:
Make pretty-lines

JNeiger reviewed (approved):
LGTM

oswinso commented on soccer/Processor.cpp:
```diff
@@ -121,6 +135,17 @@ Processor::~Processor() {
+    // when Processor dies. That normally isn't a problem, but in unit tests,
+    // we create and destroy multiple instances of Processor for each test.
+    robot_config_init = std::move(_context.robot_config);
+    std::cout << robot_config_init.get() << std::endl;
```
debug?

oswinso commented on soccer/motion/MotionSetpoint.hpp:
```diff
@@ -1,9 +1,9 @@
-    float xvelocity;
-    float yvelocity;
-    float avelocity;
+    double xvelocity;
+    double yvelocity;
+    double avelocity;
```
is there a reason for this change? Aren't floats used everywhere though?

kylestach replied:
I mean, yes they are in a lot of places, but that doesn't really mean they should be. There's really no reason not to use doubles. I changed some motion control code to use double, so I figured I might as well change this too.

oswinso replied:
I guess we should make an issue to just straight up change all floats to doubles then

kylestach replied:
Probably

oswinso reviewed (commented):
two comments

kylestach committed:
Removed debug prints

kylestach committed:
Merge branch 'staging' of github.com:robojackets/robocup-software into motion-control-node

oswinso reviewed (approved):
lgtm

kylestach committed:
Merge branch 'staging' of github.com:robojackets/robocup-software into motion-control-node
